compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1566,29 +1568,85 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
                         Some(Vec::from(pattern_spans))
                     }
                     // e.g. `let _ = Enum::TupleVariant(field1, field2);`
-                    _ if source.is_call() => {
+                    PathSource::Expr(Some(Expr { kind: ExprKind::Call(_, ref args), .. })) => {
                         err.set_primary_message(
                             "cannot initialize a tuple struct which contains private fields",
                         );
-                        if !def_id.is_local()
-                            && self
+                        if !def_id.is_local() {
+                            // Look at all the associated functions without receivers in the type's
+                            // inherent impls to look for builders that return `Self`
+                            let mut items = self
                                 .r
                                 .tcx
                                 .inherent_impls(def_id)
```

Comment:
Would be nice to also check if `Default` is implemented maybe...

Reply:
I'd rather we left that for a follow up PR :)

Reply:
Yes, I would also prefer this as a follow up =)

Reply:
May have gone overboard 👀 

Kept each commit manageable though.